File: app/src/main/java/com/nextcloud/talk/jobs/AccountRemovalWorker.java
/*
 * Nextcloud Talk - Android Client
 *
 * SPDX-FileCopyrightText: 2022 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2017 Mario Danic <mario@lovelyhq.com>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package com.nextcloud.talk.jobs;

import android.app.NotificationManager;
import android.content.Context;
import android.util.Log;

import com.nextcloud.talk.R;
import com.nextcloud.talk.api.NcApi;
import com.nextcloud.talk.application.NextcloudTalkApplication;
import com.nextcloud.talk.arbitrarystorage.ArbitraryStorageManager;
import com.nextcloud.talk.data.database.dao.ChatBlocksDao;
import com.nextcloud.talk.data.database.dao.ChatMessagesDao;
import com.nextcloud.talk.data.database.dao.ConversationsDao;
import com.nextcloud.talk.data.user.model.User;
import com.nextcloud.talk.models.json.generic.GenericMeta;
import com.nextcloud.talk.models.json.generic.GenericOverall;
import com.nextcloud.talk.models.json.push.PushConfigurationState;
import com.nextcloud.talk.users.UserManager;
import com.nextcloud.talk.utils.ApiUtils;
import com.nextcloud.talk.utils.preferences.AppPreferences;
import com.nextcloud.talk.webrtc.WebSocketConnectionHelper;

import java.net.CookieManager;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.zip.CRC32;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import autodagger.AutoInjector;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import okhttp3.JavaNetCookieJar;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;

@AutoInjector(NextcloudTalkApplication.class)
public class AccountRemovalWorker extends Worker {
    public static final String TAG = "AccountRemovalWorker";

    @Inject UserManager userManager;

    @Inject ArbitraryStorageManager arbitraryStorageManager;

    @Inject AppPreferences appPreferences;

    @Inject Retrofit retrofit;

    @Inject OkHttpClient okHttpClient;

    @Inject ChatMessagesDao chatMessagesDao;

    @Inject ConversationsDao conversationsDao;

    @Inject ChatBlocksDao chatBlocksDao;

    NcApi ncApi;

    public AccountRemovalWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        Objects.requireNonNull(NextcloudTalkApplication.Companion.getSharedApplication()).getComponentApplication().inject(this);

        List<User> users = userManager.getUsersScheduledForDeletion().blockingGet();
        for (User user : users) {
            if (user.getPushConfigurationState() != null) {
                PushConfigurationState finalPushConfigurationState = user.getPushConfigurationState();

                ncApi = retrofit
                    .newBuilder()
                    .client(okHttpClient
                                .newBuilder()
                                .cookieJar(new JavaNetCookieJar(new CookieManager()))
                                .build())
                    .build()
                    .create(NcApi.class);

                ncApi.unregisterDeviceForNotificationsWithNextcloud(
                        ApiUtils.getCredentials(user.getUsername(), user.getToken()),
                        ApiUtils.getUrlNextcloudPush(Objects.requireNonNull(user.getBaseUrl())))
                    .blockingSubscribe(new Observer<GenericOverall>() {
                        @Override
                        public void onSubscribe(@io.reactivex.annotations.NonNull Disposable d) {
                            // unused atm
                        }

                        @Override
                        public void onNext(@io.reactivex.annotations.NonNull GenericOverall genericOverall) {
                            GenericMeta meta = Objects.requireNonNull(genericOverall.getOcs()).getMeta();
                            int statusCode = Objects.requireNonNull(meta).getStatusCode();

                            if (statusCode == 200 || statusCode == 202) {
                                HashMap<String, String> queryMap = new HashMap<>();
                                queryMap.put("deviceIdentifier",
                                             finalPushConfigurationState.getDeviceIdentifier());
                                queryMap.put("userPublicKey", finalPushConfigurationState.getUserPublicKey());
                                queryMap.put("deviceIdentifierSignature",
                                             finalPushConfigurationState.getDeviceIdentifierSignature());
                                unregisterDeviceForNotificationWithProxy(queryMap, user);
                            }
                        }

                        @Override
                        public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                            Log.e(TAG, "error while trying to unregister Device For Notifications", e);
                            initiateUserDeletion(user);
                        }

                        @Override
                        public void onComplete() {
                            // unused atm
                        }
                    });
            } else {
                initiateUserDeletion(user);
            }
        }

        return Result.success();
    }

    private void unregisterDeviceForNotificationWithProxy(HashMap<String, String> queryMap, User user) {
        ncApi.unregisterDeviceForNotificationsWithProxy
                (ApiUtils.getUrlPushProxy(), queryMap)
                .subscribe(new Observer<Void>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        // unused atm
                    }

                    @Override
                    public void onNext(Void aVoid) {
                            String groupName = String.format(
                                getApplicationContext()
                                    .getResources()
                                    .getString(R.string.nc_notification_channel), user.getUserId(), user.getBaseUrl());
                            CRC32 crc32 = new CRC32();
                            crc32.update(groupName.getBytes());
                            NotificationManager notificationManager =
                                    (NotificationManager) getApplicationContext()
                                        .getSystemService(Context.NOTIFICATION_SERVICE);

                            if (notificationManager != null) {
                                notificationManager.deleteNotificationChannelGroup(
                                    Long.toString(crc32.getValue()));
                            }

                        initiateUserDeletion(user);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(TAG, "error while trying to unregister Device For Notification With Proxy", e);
                        initiateUserDeletion(user);
                    }

                    @Override
                    public void onComplete() {
                        // unused atm
                    }
                });
    }

    private void initiateUserDeletion(User user) {
        if (user.getId() != null) {
            long id = user.getId();
            WebSocketConnectionHelper.deleteExternalSignalingInstanceForUserEntity(id);

            try {
                arbitraryStorageManager.deleteAllEntriesForAccountIdentifier(id);
                deleteUser(user);
            } catch (Throwable e) {
                Log.e(TAG, "error while trying to delete All Entries For Account Identifier", e);
            }
        }
    }

    private void deleteUser(User user) {
        if (user.getId() != null) {
            String username = user.getUsername();
            try {
                userManager.deleteUser(user.getId());
                if (username != null) {
                    Log.d(TAG, "deleted user: " + username);
                }
            } catch (Throwable e) {
                Log.e(TAG, "error while trying to delete user", e);
            }
        }
    }
}

File: app/src/main/java/com/nextcloud/talk/jobs/AccountRemovalWorker.kt
/*
 * Nextcloud Talk - Android Client
 *
 * SPDX-FileCopyrightText: 2022 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2017 Mario Danic <mario@lovelyhq.com>
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package com.nextcloud.talk.jobs

import android.app.NotificationManager
import android.content.Context
import android.util.Log
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import autodagger.AutoInjector
import com.nextcloud.talk.R
import com.nextcloud.talk.api.NcApi
import com.nextcloud.talk.application.NextcloudTalkApplication
import com.nextcloud.talk.application.NextcloudTalkApplication.Companion.sharedApplication
import com.nextcloud.talk.arbitrarystorage.ArbitraryStorageManager
import com.nextcloud.talk.conversationlist.DirectShareHelper
import com.nextcloud.talk.data.database.dao.ChatBlocksDao
import com.nextcloud.talk.data.database.dao.ChatMessagesDao
import com.nextcloud.talk.data.database.dao.ConversationsDao
import com.nextcloud.talk.data.user.model.User
import com.nextcloud.talk.logger.Logger
import com.nextcloud.talk.models.json.push.PushConfigurationState
import com.nextcloud.talk.users.UserManager
import com.nextcloud.talk.utils.ApiUtils
import com.nextcloud.talk.utils.preferences.AppPreferences
import com.nextcloud.talk.webrtc.WebSocketConnectionHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.JavaNetCookieJar
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import java.net.CookieManager
import java.util.zip.CRC32
import javax.inject.Inject

@AutoInjector(NextcloudTalkApplication::class)
class AccountRemovalWorker(context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {

    @Inject
    lateinit var userManager: UserManager

    @Inject
    lateinit var arbitraryStorageManager: ArbitraryStorageManager

    @Inject
    lateinit var appPreferences: AppPreferences

    @Inject
    lateinit var retrofit: Retrofit

    @Inject
    lateinit var okHttpClient: OkHttpClient

    @Inject
    lateinit var chatMessagesDao: ChatMessagesDao

    @Inject
    lateinit var conversationsDao: ConversationsDao

    @Inject
    lateinit var chatBlocksDao: ChatBlocksDao

    @Inject
    lateinit var logger: Logger

    private lateinit var ncApi: NcApi

    override suspend fun doWork(): Result {
        sharedApplication!!.componentApplication.inject(this)

        ncApi = retrofit
            .newBuilder()
            .client(
                okHttpClient
                    .newBuilder()
                    .cookieJar(JavaNetCookieJar(CookieManager()))
                    .build()
            )
            .build()
            .create(NcApi::class.java)

        withContext(Dispatchers.IO) {
            for (user in userManager.getUsersScheduledForDeletion()) {
                val pushConfigurationState = user.pushConfigurationState
                if (pushConfigurationState != null) {
                    unregisterDeviceForNotifications(user, pushConfigurationState)
                } else {
                    initiateUserDeletion(user)
                }
            }
        }

        return Result.success()
    }

    @Suppress("TooGenericExceptionCaught")
    private suspend fun unregisterDeviceForNotifications(user: User, pushConfigurationState: PushConfigurationState) {
        val genericOverall = try {
            ncApi.unregisterDeviceForNotificationsWithNextcloud(
                ApiUtils.getCredentials(user.username, user.token),
                ApiUtils.getUrlNextcloudPush(user.baseUrl!!)
            ).blockingFirst()
        } catch (e: Exception) {
            Log.e(TAG, "error while trying to unregister Device For Notifications", e)
            initiateUserDeletion(user)
            return
        }

        val statusCode = genericOverall.ocs!!.meta!!.statusCode
        if (statusCode == HTTP_OK || statusCode == HTTP_ACCEPTED) {
            val queryMap = hashMapOf(
                "deviceIdentifier" to pushConfigurationState.deviceIdentifier,
                "userPublicKey" to pushConfigurationState.userPublicKey,
                "deviceIdentifierSignature" to pushConfigurationState.deviceIdentifierSignature
            )
            unregisterDeviceForNotificationWithProxy(queryMap, user)
        }
    }

    @Suppress("TooGenericExceptionCaught")
    private suspend fun unregisterDeviceForNotificationWithProxy(queryMap: HashMap<String, String?>, user: User) {
        try {
            ncApi.unregisterDeviceForNotificationsWithProxy(ApiUtils.getUrlPushProxy(), queryMap)
                .ignoreElements()
                .blockingAwait()

            val groupName = String.format(
                applicationContext.resources.getString(R.string.nc_notification_channel),
                user.userId,
                user.baseUrl
            )
            val crc32 = CRC32()
            crc32.update(groupName.toByteArray())
            val notificationManager =
                applicationContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
            notificationManager?.deleteNotificationChannelGroup(crc32.value.toString())
        } catch (e: Exception) {
            Log.e(TAG, "error while trying to unregister Device For Notification With Proxy", e)
        }
        initiateUserDeletion(user)
    }

    @Suppress("TooGenericExceptionCaught")
    private suspend fun initiateUserDeletion(user: User) {
        val id = user.id ?: return
        DirectShareHelper.removeShortcutsForUser(applicationContext, id)
        WebSocketConnectionHelper.deleteExternalSignalingInstanceForUserEntity(id)

        try {
            arbitraryStorageManager.deleteAllEntriesForAccountIdentifier(id)
            deleteUser(user)
        } catch (e: Exception) {
            Log.e(TAG, "error while trying to delete All Entries For Account Identifier", e)
        }
    }

    @Suppress("TooGenericExceptionCaught")
    private suspend fun deleteUser(user: User) {
        val id = user.id ?: return
        try {
            userManager.deleteUser(id)
            user.username?.let { Log.d(TAG, "deleted user: $it") }
        } catch (e: Exception) {
            Log.e(TAG, "error while trying to delete user", e)
        }
    }

    companion object {
        const val TAG = "AccountRemovalWorker"
        private const val HTTP_OK = 200
        private const val HTTP_ACCEPTED = 202
    }
}
